<a href="https://colab.research.google.com/github/GonzaloRomeroR/CEIA/blob/master/vision_computadora_2/clase_1/VP2_clase1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Ejercicio 1: Completar la tabla

In [ ]:
# Import the libraries
import torch
from torch import nn


Se obtienen las dimensiones de las salidas y los parámetros entrenables de cada una de las capas propuestas. Se imprimen estos valores por cada capa para posteriormente completar la tabla con los valores obtenidos. 

In [ ]:
# Input values 
input = torch.randn(1, 3, 64, 64)

#### Conv2D(f=5, s=2, c=8, p=2)

In [ ]:
conv_1 = nn.Conv2d(in_channels=3, out_channels=8, kernel_size=5, stride=2, padding=2)
out_conv_1 = conv_1(input)
print("Size:", out_conv_1.size())
conv_1_total_params = sum(p.numel() for p in conv_1.parameters())
print("Parameters:", conv_1_total_params)


Size: torch.Size([1, 8, 32, 32])
Parameters: 608


#### Maxpool(f=2, s=2)

In [ ]:
# Maxpool(f=2, s=2)
maxpool_1 = nn.MaxPool2d(kernel_size=2, stride=2)
out_maxpool_1 = maxpool_1(out_conv_1)
print("Size:", out_maxpool_1.size())
maxpool_1_total_params = sum(p.numel() for p in maxpool_1.parameters())
print("Parameters:", maxpool_1_total_params)

Size: torch.Size([1, 8, 16, 16])
Parameters: 0


#### Conv2D(f=3, s=1, c=16, p=1)

In [ ]:
conv_2 = nn.Conv2d(in_channels=8, out_channels=16, kernel_size=3, stride=1, padding=1)
out_conv_2 = conv_2(out_maxpool_1)
print("Size:", out_conv_2.size())
conv_2_total_params = sum(p.numel() for p in conv_2.parameters())
print("Parameters:", conv_2_total_params)


Size: torch.Size([1, 16, 16, 16])
Parameters: 1168


#### Maxpool(f=4, s=4)

In [ ]:
# Maxpool(f=4, s=4)
maxpool_2 = nn.MaxPool2d(kernel_size=4, stride=4)
out_maxpool_2 = maxpool_2(out_conv_2)
print("Size:", out_maxpool_2.size())
maxpool_2_total_params = sum(p.numel() for p in maxpool_2.parameters())
print("Parameters:", maxpool_2_total_params)

Size: torch.Size([1, 16, 4, 4])
Parameters: 0


#### Dense(salida=6)

In [ ]:
# Dense(salida=6)
linear = nn.Linear(in_features=16 * 4 * 4, out_features=6)
out_linear = linear(out_maxpool_2.view(-1, 16 * 4 * 4))
print("Size:", out_linear.size())
linear_total_params = sum(p.numel() for p in linear.parameters())
print("Parameters:", linear_total_params)


Size: torch.Size([1, 6])
Parameters: 1542


### Tabla



|       | Dimensiones de entrada | Cantidad de parámetros |
| ----------- | ----------- | ----------- |
| Entrada      | (3, 64, 64)       | 0       |
| Conv2D(f=5, s=2, c=8, p=2)  | (8, 32, 32)        | 608        |
| MaxPool(f=2, s=2)  | (8, 16, 16)        | 608        |
| Conv2D(f=3, s=1, c=16, p=1)  | (16, 16, 16)        | 1168        |
| MaxPool(f=4, s=4)  | (16, 4, 4)        | 0        |
| Dense(salida=6)  | (6)        | 1542        | 